src/Identity/Core/src/DefaultPasskeyHandler.cs

```diff
@@ -0,0 +1,602 @@
+        //        Handles all validation related to the attestation statement (21-24).
+        var isAttestationStatementValid = await VerifyAttestationStatementAsync(attestationObjectMemory, clientDataHash, context.HttpContext).ConfigureAwait(false);
+        if (!isAttestationStatementValid)
+        {
+            throw PasskeyException.InvalidAttestationStatement();
+        }
+
+        // 25. Verify that the credentialId is <= 1023 bytes.
+        // NOTE: Handled while parsing the attested credential data.
+        if (!credential.Id.AsSpan().SequenceEqual(attestedCredentialData.CredentialId.Span))
+        {
+            throw PasskeyException.CredentialIdMismatch();
+        }
+
+        var credentialId = attestedCredentialData.CredentialId.ToArray();
+
+        // 26. Verify that the credentialId is not yet registered for any user.
+        var existingUser = await context.UserManager.FindByPasskeyIdAsync(credentialId).ConfigureAwait(false);
+        if (existingUser is not null)
+        {
+            throw PasskeyException.CredentialAlreadyRegistered();
+        }
+
+        // 27. Let credentialRecord be a new credential record with the following contents:
+        var credentialRecord = new UserPasskeyInfo(
+            credentialId,
+            publicKey: attestedCredentialData.CredentialPublicKey.ToArray(),
+            name: null,
+            createdAt: DateTime.UtcNow,
+            signCount: authenticatorData.SignCount,
+            transports: response.Transports,
+            isUserVerified: authenticatorData.IsUserVerified,
+            isBackupEligible: authenticatorData.IsBackupEligible,
+            isBackedUp: authenticatorData.IsBackedUp,
+            attestationObject: response.AttestationObject.ToArray(),
+            clientDataJson: response.ClientDataJSON.ToArray());
+
+        // 28. Process the client extension outputs in clientExtensionResults and the authenticator extension
+        //     outputs in the extensions in authData as required by the Relying Party.
+        // NOTE: Not currently supported.
+
+        // 29. If all the above steps are successful, store credentialRecord in the user account that was denoted
+        // and continue the registration ceremony as appropriate.
+        return PasskeyAttestationResult.Success(credentialRecord);
+    }
+
+    /// <summary>
+    /// Performs passkey assertion using the provided credential JSON, original options JSON, and optional user.
+    /// </summary>
+    /// <param name="context">The context containing necessary information for passkey assertion.</param>
+    /// <returns>A task object representing the asynchronous operation containing the <see cref="PasskeyAssertionResult{TUser}"/>.</returns>
+    protected virtual async Task<PasskeyAssertionResult<TUser>> PerformAssertionCoreAsync(PasskeyAssertionContext<TUser> context)
+    {
+        // See https://www.w3.org/TR/webauthn-3/#sctn-registering-a-new-credential
```

Comment:
Is this link correct?

Should this be https://www.w3.org/TR/webauthn-3/#sctn-verifying-assertion instead?

Reply:
Yes, it should be that. Good catch. Thanks 🙂